File: 1080.insufficient-nodes-in-root-to-leaf-paths/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* sufficientSubset(TreeNode* root, int limit) {
        if (nullptr == root)
            return nullptr;
        if (root->left == nullptr && root->right == nullptr) {
            if (root->val < limit) {
                return nullptr;
            }
            return root;
        }
        if (root->left != nullptr) root->left = sufficientSubset(root->left, limit - root->val);
        if (root->right != nullptr) root->right = sufficientSubset(root->right, limit - root->val);
        if (root->left == nullptr && root->right == nullptr) return nullptr;
        return root;
    }
};

File: 1080.insufficient-nodes-in-root-to-leaf-paths/solution.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun sufficientSubset(root: TreeNode?, limit: Int): TreeNode? {
        fun checkSufficientLeaf(node: TreeNode?, sum: Int, limit: Int): Boolean {
            if (node == null) return false
            if (node.left == null && node.right == null) return node.`val` + sum >= limit
            val haveSufficientLeft = checkSufficientLeaf(node.left, sum + node.`val`, limit)
            val haveSufficientRight = checkSufficientLeaf(node.right, sum + node.`val`, limit)
            if (!haveSufficientLeft) node.left = null
            if (!haveSufficientRight) node.right = null
            return haveSufficientLeft || haveSufficientRight
        }

        return if(checkSufficientLeaf(root, 0, limit)) root else null
    }
}
